dali/kernels/scratch_copy_impl.h

```diff
@@ -0,0 +1,171 @@
+///          no deletion.
+template <typename AlignedAllocator, typename... Collections,
+          typename = decltype(std::declval<AlignedAllocator>()(0, 0))>
+auto ToContiguousHostMem(AlignedAllocator &alloc, const Collections &... c) {
+  const size_t N = sizeof...(Collections);
+  static_assert(
+    all_of<std::is_pod<std::remove_cv_t<element_t<Collections>>>::value...>::value,
+    "ToGPU must be used with collections of POD types");
+
+  std::array<size_t, N + 1> offsets;
+  detail::GetCollectionOffsets(0, &offsets[0], c...);
+  size_t alignment = detail::variadic_max(alignof(element_t<Collections>)...);
+  size_t total_size = std::get<N>(offsets);
+
+  auto *tmp = reinterpret_cast<char*>(alloc(total_size, alignment));
+  detail::copy_to_buffer(tmp, &offsets[0], c...);
+
+  return detail::GetCollectionPtrs(tmp, &offsets[0], c...);
+}
+
+/// @brief Allocates GPU memory using provided allocator, copies the collections to a
+///        temporary host buffer and then transfers the contents to the GPU in just one
+///        `cudaMemcpyAsync`.
+///
+/// @remarks The first pointer in the output tuple is the one returned by the allocator.
+///          Note that for memory returned by PreallocatedScratchpad, this pointer needs
+///          no deletion.
+template <typename AlignedAllocator, typename... Collections,
+          typename = decltype(std::declval<AlignedAllocator>()(0, 0))>
+auto ToContiguousGPUMem(AlignedAllocator &&alloc, cudaStream_t stream, const Collections &... c) {
+  const size_t N = sizeof...(Collections);
+  static_assert(
+    all_of<std::is_pod<std::remove_cv_t<element_t<Collections>>>::value...>::value,
+    "ToGPU must be used with collections of POD types");
+
+  std::array<size_t, N + 1> offsets;
+  detail::GetCollectionOffsets(0, &offsets[0], c...);
+  size_t alignment = detail::variadic_max(alignof(element_t<Collections>)...);
+  size_t total_size = std::get<N>(offsets);
+
+  char *tmp;
+  std::unique_ptr<char[]> heap_buf;
+  if (total_size <= 0x2000) {
+     tmp = static_cast<char*>(alloca(0x2000));
+  } else {
+    heap_buf.reset(new char[total_size]);
+    tmp = heap_buf.get();
+  }
+
+  detail::copy_to_buffer(tmp, &offsets[0], c...);
+  void *out_ptr = alloc(total_size, alignment);
+
+  cudaMemcpyAsync(out_ptr, tmp, total_size, cudaMemcpyHostToDevice, stream);
+  return detail::GetCollectionPtrs(out_ptr, &offsets[0], c...);
+}
+
+template <typename... Collections>
+std::tuple<std::remove_cv_t<element_t<Collections>>*...>
+ToContiguousHostMem(Scratchpad &scratchpad, const Collections &... c) {
+  auto alloc = [&](size_t n, size_t align) {
+    return scratchpad.Alloc(AllocType::Host, n, align);
+  };
+  return ToContiguousHostMem(alloc, c...);
+}
+
+template <typename... Collections>
+std::tuple<std::remove_cv_t<element_t<Collections>>*...>
+ToContiguousGPUMem(Scratchpad &scratchpad, cudaStream_t stream, const Collections &... c) {
```

Comment:
Can this two `ToContiguousGPUMem` be merged into one function? What is the benefit of having them separated?
Same for `ToContiguousHostMem`.

Reply:
At least in theory, the larger function can be used with arbitrary allocator capable of providing memory in appropriate address space. This one is a convenient wrapper for Scratchpad, because that's where it will most likely be used.

Reply:
But do we have any other allocator we want to put it or it is just overgeneralization?

Reply:
Not at present.

Reply:
Removed.